src/abi/anchor.rs

```diff
@@ -17,3 +28,375 @@ pub fn discriminator(namespace: &'static str, name: &str) -> Vec<u8> {
+            func.name.clone()
+        };
+
+        let returns = if func.returns.is_empty() {
+            None
+        } else if func.returns.len() == 1 {
+            Some(type_manager.convert(&func.returns[0].ty))
+        } else {
+            Some(type_manager.build_struct_for_return(func, &name))
+        };
+
+        instructions.push(IdlInstruction {
+            name,
+            docs: tags,
+            accounts,
+            args,
+            returns,
+        });
+    }
+
+    instructions
+}
+
+/// This struct accounts all the user defined types used in the contract that need to be present
+/// in the IDL 'types' field.
+struct TypeManager<'a> {
+    namespace: &'a Namespace,
+    added_types: HashSet<Type>,
+    returns_structs: Vec<IdlTypeDefinition>,
+    types: Vec<IdlTypeDefinition>,
+}
+
+impl TypeManager<'_> {
+    fn new(ns: &Namespace) -> TypeManager {
+        TypeManager {
+            namespace: ns,
+            added_types: HashSet::new(),
+            types: Vec::new(),
+            returns_structs: Vec::new(),
+        }
+    }
+
+    /// Functions with multiple returns must return a struct in Anchor, so we build a return
+    /// struct containing all the returned types.
+    fn build_struct_for_return(&mut self, func: &Function, effective_name: &String) -> IdlType {
+        let mut fields: Vec<IdlField> = Vec::with_capacity(func.returns.len());
+        for (item_no, item) in func.returns.iter().enumerate() {
+            let name = if let Some(id) = &item.id {
+                id.name.clone()
+            } else {
+                format!("return_{}", item_no)
+            };
+
+            fields.push(IdlField {
+                name,
+                docs: None,
+                ty: self.convert(&item.ty),
+            });
+        }
+
+        let name = format!("{}_returns", effective_name);
+        self.returns_structs.push(IdlTypeDefinition {
+            name: name.clone(),
+            docs: Some(vec![format!(
+                "Data structure to hold the multiple returns of function {}",
+                func.name
+            )]),
+            ty: IdlTypeDefinitionTy::Struct { fields },
+        });
+
+        IdlType::Defined(name)
+    }
+
+    /// Add a struct definition to the TypeManager
+    fn add_struct_definition(&mut self, def: &StructDecl, ty: &Type) {
+        if self.added_types.contains(ty) {
+            return;
+        }
+        self.added_types.insert(ty.clone());
+
+        let docs = idl_docs(&def.tags);
+
+        let mut fields: Vec<IdlField> = Vec::with_capacity(def.fields.len());
+        for item in &def.fields {
+            fields.push(IdlField {
+                name: item.name_as_str().to_string(),
+                docs: None,
+                ty: self.convert(&item.ty),
+            });
+        }
+
+        self.types.push(IdlTypeDefinition {
+            name: def.name.clone(),
+            docs,
+            ty: IdlTypeDefinitionTy::Struct { fields },
+        });
```

Comment:
I just realized there is an issue.

```solidity
contract D {
	struct Foo { int64 f1; }
}
contract C {
	enum Foo { b1, b2, b3 }
}
```
Now we have two types with the same name `Foo`.

Reply:
There is a different IDL file for each contract.

Reply:
In addition, custom types only appear in the IDL if they part of a function argument or return.

Reply:
```solidity
contract D {
	struct Foo { int64 f1; }
}
contract C {
	enum Foo { b1, b2, b3 }
        function f(D.Foo x) public returns (int64) { return x.f1; }
}